code/game/objects/items/glassjar.dm

```diff
@@ -31,75 +33,109 @@
 /obj/item/glass_jar/MouseDrop(atom/target)
 	if(ishuman(target) || issmall(target) && target == usr && use_check_and_message(usr) && Adjacent(usr))
 		if(contains == JAR_GUMBALL)
-			handle_gumball_removal(usr)
+			if(length(gumballs_contained))
+				usr.put_in_hands(gumballs_contained[1])
+				gumballs_contained -= gumballs_contained[1]
+				usr.visible_message("<b>[usr]</b> takes a gumball from \the [src].", SPAN_NOTICE("You take a gumball from \the [src]."))
+				if(length(gumballs_contained) == 0)
+					contains = JAR_NOTHING
+				update_icon()
+
 
 /obj/item/glass_jar/afterattack(var/atom/A, var/mob/user, var/proximity)
 	if(!proximity || contains)
 		return
 	if(istype(A, /obj/effect/spider/spiderling))
 		var/obj/effect/spider/spiderling/S = A
-		user.visible_message(SPAN_NOTICE("<b>\The [user]</b> scoops \the [S] into \the [src]."), SPAN_NOTICE("You scoop \the [S] into \the [src]."))
-		S.forceMove(src)
 		STOP_PROCESSING(SSprocessing, S)	// No growing inside jars
 		contains = JAR_SPIDERLING
-		update_icon()
-		return
+		scoop(S, user)
 	if(ismob(A))
 		var/mob/L = A
 		if(L.mob_size <= MOB_TINY)
-			user.visible_message(SPAN_NOTICE("<b>\The [user]</b> scoops \the [L] into \the [src]."), SPAN_NOTICE("You scoop \the [L] into \the [src]."))
-			L.forceMove(src)
 			contains = JAR_ANIMAL
-			update_icon()
-			return
+			scoop(L, user)
 		else
-			to_chat(user, SPAN_WARNING("\The [L] doesn't fit into \the [src]!"))
-			return
+			scoop_fail(L, user)
+
+/obj/item/glass_jar/proc/scoop(var/atom/movable/A, var/mob/user)
+	user.visible_message(SPAN_NOTICE("<b>[user]</b> scoops \the [A] into \the [src]."), SPAN_NOTICE("You scoop \the [A] into \the [src]."))
+	playsound(src, pickup_sound, PICKUP_SOUND_VOLUME)
+	A.forceMove(src)
+	update_icon()
+	return
+
+/obj/item/glass_jar/proc/scoop_fail(var/atom/A, var/mob/user)
+	to_chat(user, SPAN_WARNING("\The [A] doesn't fit into \the [initial(name)]!"))
+	playsound(src, drop_sound, DROP_SOUND_VOLUME)
+	return
 
 /obj/item/glass_jar/attack_self(var/mob/user)
 	switch(contains)
 		if(JAR_MONEY)
 			for(var/obj/O in src)
 				O.forceMove(user.loc)
-			to_chat(user, SPAN_NOTICE("You take money out of \the [src]."))
-			contains = JAR_NOTHING
-			update_icon()
-			return
+				release(O, user)
```

Comment:
you should probably only use release() for things that can't be held in hand - 
taking money or gumballs or w/e out of a jar should try to put them in the user's active hand